README.md
# Fire Paystack

This repo contains the implementation for the Paystack Firebase Extension

## How to use this extension

Use this extension to process payments in Africa with [Paystack](https://paystack.com) based on the information contained in a document added to a specific Firestore collection.

Adding a document triggers this extension to [initialise a payment](https://paystack.com/docs/api/transaction/#initialize) with Paystack, and writes an `authorization_url back` to the document. This `authroization_url` can then be used to direct a user to the payment screen hosted by Paystack. The document’s top-level fields specify the email of the purchaser and the amount of the payment. Currently, payments are only supported in ZAR, but expanded support for multiple currencies is planned for the future.

Here’s a basic example document write that would trigger this extension to initialize a payment of R100 for `someone@example.com`:

```js
admin.firestore().collection('payments').add({
  email: 'someone@example.com',
  amount: 100,
})
```

This would then populate the document with some additional fields, in particular the `authorization_url` which is the payment page hosted by paystack. After successful payment, the extension has an HTTPS triggered Cloud Function which is called by Paystack to alert us of the success. This will then set the `status` field of the payment document to "successful".

## Setup with Paystack

When you configure this extension, you’ll need to supply your Paystack API key for the extension to interact with Paystack. This can be found on your Paystack Dashboard under [Settings > API Keys and Webhooks](https://dashboard.paystack.com/#/settings/developers).

In this same section of the settings, you will provide the webhook URL corresponding to the Cloud Function that has been created by the extension. Make sure to provide the URL corresponding to a devlopment project in the "Test Webhook URL" and the URL corresponding to a production project in the "Live Webhook URL."

See below the section to access the Test API key and setting of the webhook URL"
![Paystack settings](paystack-settings.png)

The webhook URL can be found by navigating to the Cloud Functions section of your Firebase project in which the extension is installed:
![Firebase Cloud Functions](cloud-functions.png)

<!-- We recommend keeping the following section to explain how billing for Firebase Extensions works -->
## Billing

To install an extension, your project must be on the Blaze (pay as you go) plan

This extension uses other Firebase and Google Cloud Platform services, which have associated charges if you exceed the service’s no-cost tier:

- Secrets Manager
- Cloud Functions
- Cloud Firestore

When you use Firebase Extensions, you're only charged for the underlying resources that you use. All Firebase services offer a free tier of usage. [Learn more about Firebase billing.](https://firebase.google.com/pricing). Furthermore, the [transaction fees](https://paystack.com/za/pricing) as charged by Pastack are applied to all payments processed.


CHANGELOG.md
## Version 0.0.1
- Initial Version
  
## Version 0.0.2
- Added origin verification to handlePayment webhook to prevent fraudulent payment confirmations
- Added additional commentary and documentation

## Version 0.0.3
- More touch ups to documentation


## Version 0.0.4
- Demo version for Google DevFest

functions/src/paystack/client.test.ts
import { PaystackClient } from "./client";
import axios from "axios";
import * as sinon from "sinon";
import * as chai from "chai";
import * as chaiAsPromised from "chai-as-promised";

chai.use(chaiAsPromised);
const expect = chai.expect;

describe("PaystackClient", () => {
  const secret = "your-secret-key";
  const client = new PaystackClient(secret);
  let axiosRequestStub: sinon.SinonStub;

  beforeEach(() => {
    axiosRequestStub = sinon.stub(axios, "request");
  });

  afterEach(() => {
    sinon.restore();
  });

  describe("initializeTransaction", () => {
    it("should send a POST request to the correct URL with the provided data", async () => {
      const request = {
        amount: 100,
        email: "jasonsbrown25@gmail.com",
        currency: "ZAR"
      };
      const expectedResponse = {
        
      };

      axiosRequestStub.resolves({ data: expectedResponse });

      const response = await client.initializeTransaction(request);

      expect(
        axiosRequestStub.calledWith({
          url: "https://api.paystack.co/transaction/initialize",
          method: "POST",
          headers: {
            Accept: "application/json",
            "Content-Type": "application/json",
            Authorization: `Bearer ${secret}`,
          },
          data: request,
        })
      ).to.be.true;

      expect(response).to.eql(expectedResponse);
    });

    it("should throw an error if the response status is not in the 200-399 range", async () => {
      const request = {
        amount: 100,
        email: "jasonsbrown25@gmail.com",
        currency: "ZAR",
      };
      const errorResponse = { status: 400, message: "Bad Request" };

      axiosRequestStub.resolves({ status: 400, data: errorResponse });

      await expect(client.initializeTransaction(request)).to.be.rejectedWith(
        "400: Bad Request"
      );
    });
  });
});


functions/src/index.ts
/*
 * This template contains a HTTP function that responds
 * with a greeting when called
 *
 * Reference PARAMETERS in your functions code with:
 * `process.env.<parameter-name>`
 * Learn more about building extensions in the docs:
 * https://firebase.google.com/docs/extensions/publishers
 */

import { PaystackClient } from "./paystack/client";
import { InitializeTransactionResponse, PaystackEvent } from "./paystack/types";

import { firestore } from "firebase-functions/v1";
import { initializeApp } from "firebase-admin/app";
import { getFirestore } from "firebase-admin/firestore";
import * as functions from "firebase-functions";
import * as crypto from "crypto";

initializeApp();

// Initialise variables
const secret = process.env.PAYSTACK_API_KEY || "";
const paymentsCollection = process.env.PAYMENTS_COLLECTION || "payments";
const defaultCurrency = process.env.DEFAULT_CURRENCY || "ZAR";

const paystackClient = new PaystackClient(secret);

// paymentCreated is triggers when a document is added to the payments collection.
// The information in this document is used to initialize a payment with Paystack
// and the corresponding authoization URL for the payment is written back to 
// the document
export const paymentCreated = firestore
  .document(`${paymentsCollection}/{paymentID}`)
  .onCreate(async (snapshot, context) => {
    console.log("Payment creeated");

    // In order to create a payment, we need the following information:
    // 1. The email of the buyer
    // 2. The amount of the purchase
    // 3. The currency
    const paymentDetails = snapshot.data();

    // Check that the payment has the necessary fields and initialize the transaction
    if (!paymentDetails || !paymentDetails.email || !paymentDetails.amount) {
      // Write an error to the document if there is a problem initializing the transaction
      try {
        await snapshot.ref.set(
          {
            error: "Must have email and amount set",
          },
          { merge: true }
        );
      } catch (error) {
        console.error("Failed to write error to payment document", error);
      }
      return;
    }

    // Create the transaction request
    const transactionRequest = {
      email: paymentDetails.email,
      amount: paymentDetails.amount * 100,
      currency: defaultCurrency,
      // The reference is the unique ID assigned to payments as tracked in your backend
      reference: context.params.paymentID,
    };

    let transactionResponse: InitializeTransactionResponse;
    try {
      // Initialize the transaction
      transactionResponse = await paystackClient.initializeTransaction(
        transactionRequest
      );
      console.log("Transaction initialized", transactionResponse);

      // Write the redirect authorization URL response data back to the payment
      // for use by clients
      await snapshot.ref.set(transactionResponse, { merge: true });
    } catch (error) {
      console.error("Failed to initialize transaction", error);
    }
  });

// handlePayment is the webhook that Paystack will send events to
// to acknowledge receirpt of payment. We will use this event
// to set the status of the payment accordingly
export const handlePayment = functions
  .runWith({
    // Ensure the function has enough memory and time
    // to process large files
    timeoutSeconds: 300,
  })
  .https.onRequest(async (req, res) => {
    // Verify the origin
    const hash = crypto
      .createHmac("sha512", secret)
      .update(JSON.stringify(req.body))
      .digest("hex");
    if (hash != req.headers["x-paystack-signature"]) {
      // The origin is not from Paystack, warn and return error
      console.warn("Origin is not Paystack. Denying verification");
      res.status(403).send("Access denied");
    }

    try {
      const event: PaystackEvent = req.body;

      switch (event.event) {
        case "charge.success":
          // Assuming that we store the payment details in a collection 'payments'
          // and the Paystack transaction ID is used as the document ID in Firestore
          const paymentRef = getFirestore()
            .collection(paymentsCollection)
            .doc(event.data.reference.toString());

          // Update the payment document with the successful charge details
          await paymentRef.set(
            {
              status: "successful",
              gateway_response: event.data.gateway_response,
            },
            { merge: true }
          );

          break;

        default:
          // Handle unhandled events
          console.warn("Unhandled Paystack event: ", event.event);
          break;
      }

      res.status(200).send("Event processed");
    } catch (error) {
      console.error("Error handling Paystack event: ", error);
      res.status(500).send("Internal Server Error");
    }
  });


functions/src/paystack/types.ts
/**
 * Interface for the request to initialize a transaction.
 */
export interface InitializeTransactionRequest {
  amount: number, // The amount for the transaction
  email: string, // The email of the user initiating the transaction
  currency: string, // The currency for the transaction
  reference?: string, // Optional reference for the transaction
  callback_url?: string, // Optional callback URL for the transaction
}

/**
 * Interface for the response after initializing a transaction.
 */
export interface InitializeTransactionResponse {
  authorization_url: string; // The authorization URL for the transaction
  access_code: string; // The access code for the transaction
  reference: string; // The reference for the transaction
}

export interface PaystackEventData {
  id: number;
  reference: string;
  gateway_response: string;
}

export interface PaystackEvent {
  event: string;
  data: PaystackEventData;
}

/**
 * Interface for the response from the Paystack API.
 */
export interface Response<T> {
  status: boolean; // The status of the response
  message: string; // The message from the response
  data: T; // The data from the response
}


functions/src/paystack/client.ts
import axios from "axios";
import {
  Response,
  InitializeTransactionRequest,
  InitializeTransactionResponse,
} from "./types";

const PAYSTACK_BASE_URL = "https://api.paystack.co";

/**
 * This class represents a client for the Paystack API.
 */
export class PaystackClient {
  /**
   * The secret key used for authenticating with the Paystack API.
   */
  private secret: string;

  /**
   * Creates a new PaystackClient.
   * @param {string} secret - The secret key used for authenticating
   * with the Paystack API.
   */
  constructor(secret: string) {
    this.secret = secret;
  }

  /**
   * Sends a request to the specified URL using the specified method.
   * @param {string} url - The URL to send the request to.
   * @param {"GET" | "POST"} method - The HTTP method to use for the request.
   * @param {any} data - The data to send with the request (optional).
   * @return {Promise<T>} - A promise that resolves to the response data.
   * @throws {Error} - If the response status is not in the 200-399 range.
   */
  private async sendRequest<T>(
    url: string,
    method: "GET" | "POST",
    data?: any
  ): Promise<T> {
    const headers = {
      "Accept": "application/json",
      "Content-Type": "application/json",
      "Authorization": `Bearer ${this.secret}`,
    };

    const response = await axios.request<Response<T>>({
      url: `${PAYSTACK_BASE_URL}${url}`,
      method,
      headers,
      data,
    });

    if (response.status < 200 || response.status >= 400) {
      console.log("ERROR")
      const errorResponse: Response<any> = response.data as Response<any>;
      throw new Error(`${response.status}: ${errorResponse.message}`);
    }

    return response.data.data as T;
  }

  /**
   * Initializes a transaction.
   * @param {InitializeTransactionRequest} request - The request object for
   * initializing the transaction.
   * @return {Promise<InitializeTransactionResponse>} - A promise that
   * resolves to the response object.
   */
  public async initializeTransaction(
    request: InitializeTransactionRequest
  ): Promise<InitializeTransactionResponse> {
    return await this.sendRequest<InitializeTransactionResponse>(
      "/transaction/initialize",
      "POST",
      request
    );
  }
}
